src/OneDriveSdk/Helpers/UploadSessionHelper.cs

```diff
@@ -0,0 +1,257 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.OneDrive.Sdk.Helpers
+{
+    using Microsoft.Graph;
+
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    public class UploadSessionHelper
+    {
+        private const int DefaultMaxChunkSize = 10 * 1024 * 1024;
+
+        public UploadSession Session { get; private set; }
+        private IBaseClient client;
+        private Stream uploadStream;
+        private int totalUploadLength;
+        private readonly int maxChunkSize;
+        private List<Tuple<int, int>> rangesRemaining;
+        
+        /// <summary>
+        /// Helps with resumable uploads. Generates chunk requests based on <paramref name="session"/>
+        /// information, and can control uploading of requests using <paramref name="client"/>
+        /// </summary>
+        /// <param name="session">Session information.</param>
+        /// <param name="client">Client used to upload chunks.</param>
+        /// <param name="uploadStream">Readable, seekable stream to be uploaded.</param>
+        /// <param name="totalUploadLength">Total length of file to uploaded. <paramref name="uploadStream"/> must have at least
+        /// this many bytes.</param>
+        /// <param name="maxChunkSize">Max size of each chunk to be uploaded. If less than 0, default value of 10 MB is used. Increment of 320 kb is
+        /// recommended.</param>
+        public UploadSessionHelper(UploadSession session, IBaseClient client, Stream uploadStream, int totalUploadLength, int maxChunkSize = -1)
+        {
+            if (!uploadStream.CanRead || !uploadStream.CanSeek)
+            {
+                throw new ArgumentException("Must provide stream that can read and seek");
+            }
+
+            this.Session = session;
+            this.client = client;
+            this.uploadStream = uploadStream;
+            this.totalUploadLength = totalUploadLength;
+            this.rangesRemaining = this.GetRangesRemaining(session);
+            this.maxChunkSize = maxChunkSize < 0 ? DefaultMaxChunkSize : maxChunkSize;
```

Comment:
>maxChunkSize [](start = 17, length = 12)

"Use a fragment size that is a multiple of 320 KiB (320 * 1024 bytes). Failing to use a fragment size that is a multiple of 320 KiB can result in large file transfers failing after the last fragment is uploaded."  - We should throw on chunksizes that are not divisible by 320*1024